extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/shuffle/DurableStorageInputChannelFactory.java

```diff
@@ -112,11 +119,33 @@ public ReadableFrameChannel openChannel(StageId stageId, int workerNumber, int p
     catch (Exception e) {
       throw new IOE(
           e,
-          "Could not find remote output of worker task[%s] stage[%d] partition[%d]",
-          workerTaskId,
+          "Could not find remote output of worker task[%d] stage[%d] partition[%d]",
+          workerNumber,
           stageId.getStageNumber(),
           partitionNumber
       );
     }
   }
+
+  @Nullable
+  public String findSuccessfulPartitionOutput(
+      final String controllerTaskId,
+      final int workerNo,
+      final int stageNumber,
+      final int partitionNumber
+  ) throws IOException
+  {
+    List<String> fileNames = storageConnector.lsFiles(
+        DurableStorageOutputChannelFactory.getPartitionOutputsFolderName(
+            controllerTaskId,
+            workerNo,
+            stageNumber,
+            partitionNumber
+        )
+    );
+    Optional<String> maybeFileName = fileNames.stream()
```

Comment:
So we can have issues here. 
This logic is not deterministic. 
Lets say we have a case where one zombie task also wrote its output successfully to durable storage. This means, the output of that task will also be considered to be read 
So now you have 
`/xxx/workerNumber/task_good_id/part_1_success`
`/xxx/workerNumber/task_zombie_id/part_1_success`

As rows in both of them may not follow the same order we might be in a soup if worker one reads zombie_task files and worker 2 reads good_id files. 


The best fix would be to figure out which path to read the data from at one place IE in the controller. 

If we are not able to do that then we have to make some changes into how the correct taskID files are picked. 

I suggest using time as the tiebreaker. Which ever successful file is written first, we use that. Basically, first write wins. 



cc @gianm

Reply:
> As rows in both of them may not follow the same order we might be in a soup if worker one reads zombie_task files and worker 2 reads good_id files.

I was under the assumption that the row ordering in both of the successful writes should be the same.
Considering a non shuffling case, it doesn't change the row order, therefore the output of the successful write should be identical to the input (which if we track back to the original input source, should have a fixed row order.
In a shuffling case, we might sort the rows; therefore, the output must be similar to the sort done. I think there might be some indeterminism there (which I doubt considering that the FrameChannelMerger for both workers should produce the same output), but when we read the sorted data, we pass it through the `FrameChannelMerger` so as long as the rows are identical (and not the order) in the files we shouldn't have an issue.

This is all considering that the successful files have different outputs, which I don't think should be the case 🤔. WDYT?

Reply:
Thinking about it further, even if the order is different, however, the content is the same, it shouldn't matter since a single worker is assigned to a single partition, right?

Reply:
As we are downloading segments in different threads, outputs are not guaranteed to be same. 
There can be a case where one parent of the dag is read by two child nodes and then we want to pass the same outputs to both the cases especially if we do limit/offset. Also, a deterministic system is always easier to debug.  
```
   ()
 /    \
()    ()
```

Reply:
Thanks! As per your original suggestions, I like the time-based approach. Another one that I can think of is picking up the first file in alphabetically sorted order. 
If we are using one of the above methods, then I see no inherent benefits in using the controller to figure out the file in case of a tiebreaker.

Reply:
Consider the scenario where:
1. File with task id which starts with B is written at t0
2. The same partition with task id which starts with A is written at t5

The alphabetically sorted order will give different results at t4 and t6 which is what we would want to avoid.